### ISSUE TITLE
org.openscience.cdk.libio.cml.Convertor WARN: Duplicate attempt to register a customizer

### ISSUE THREAD

**Issue by blurock** [cdk/cdk#557]
I am obviously using `Convertor` wrong... I think with each instantiation, I get this warning.
`org.openscience.cdk.libio.cml.Convertor WARN: Duplicate attempt to register a customizer`

The question is where and how I should 'register' a customer for the `Convertor`?
Or is there a way around this?

Ultimately, I want to do (several times):
```java
CMLMolecule cml = convert.cdkAtomContainerToCMLMolecule(molecule);
```

What is the 'proper' technique?

At the moment I am just doing:

```java
String molname = "Molecule";
Convertor convert = new Convertor(true, nameOfStructure);
CMLMolecule cml = convert.cdkAtomContainerToCMLMolecule(molecule);

I tried with a static instance, with the same results:
	private static Convertor standardconvertor;
	private static Convertor getConvertor() {
		if(standardconvertor == null) {
			standardconvertor = new Convertor(true,null);
		}
		return standardconvertor;
	}
```

thanks for any help...
Regards
Edward Blurock

**Comment by egonw** [comment]
I'll get back on this asap... need to check my old notes :)

**Comment by blurock** [comment]
_As an annotation to what I wrote before...
I am not how, as an outsider, can fix this problem...

Here is my partial analysis

Looking at the use of Convertor, for example, in CMLWriter (where I am mostly using it)._
        **Convertor convertor = new Convertor**(cmlIds.isSet(),
                (namespacePrefix.getSetting().length() > 0) ? namespacePrefix.getSetting() : null);

_I think it has to be defined as a static class on the order of:_

	private Convertor standardconvertor;
	private Convertor getConvertor() {
		if(standardconvertor == null) {
			System.out.println("getConverter------------------------------------------------");
			standardconvertor = new Convertor(true,null);
		}
		return standardconvertor;
	}


I am pretty sure the error is coming from (within Convertor)
   private void setupCustomizers() {
        if (customizers == null) customizers = new HashMap<String, ICMLCustomizer>();

        try {
            logger.debug("Starting loading Customizers...");
            BufferedReader reader = new BufferedReader(new InputStreamReader(this.getClass().getClassLoader()
                    .getResourceAsStream(CUSTOMIZERS_LIST)));
            int customizerCount = 0;
            while (reader.ready()) {
                // load them one by one
                String customizerName = reader.readLine();
                customizerCount++;
                if (customizers.containsKey(customizerName)) {
                    try {
                        ICMLCustomizer customizer = (ICMLCustomizer) this.getClass().getClassLoader()
                                .loadClass(customizerName).newInstance();
                        customizers.put(customizer.getClass().getName(), customizer);
                        logger.info("Loaded Customizer: ", customizer.getClass().getName());
                    } catch (ClassNotFoundException exception) {
                        logger.info("Could not find this Customizer: ", customizerName);
                        logger.debug(exception);
                    } catch (InstantiationException | IllegalAccessException exception) {
                        logger.warn("Could not load this Customizer: ", customizerName);
                        logger.warn(exception.getMessage());
                        logger.debug(exception);
                    }
                } else {
                    logger.warn("**Duplicate attempt to register a customizer**");
                }
            }
            logger.info("Number of loaded customizers: ", customizerCount);
        } catch (Exception exception) {
            logger.error("Could not load this list: ", CUSTOMIZERS_LIST);
            logger.debug(exception);
        }
    }

**Comment by johnmay** [comment]
There is an example of using a customeriser in "QSARCMLRoundTripTest".

```java
StringWriter writer = new StringWriter();
IAtomContainer molecule = TestMoleculeFactory.makeBenzene();
IMolecularDescriptor descriptor = new WeightDescriptor();

CMLWriter cmlWriter = new CMLWriter(writer);
cmlWriter.registerCustomizer(new QSARCustomizer());
DescriptorValue value = descriptor.calculate(molecule);
molecule.setProperty(value.getSpecification(), value);

cmlWriter.write(molecule);
String cmlContent = writer.toString();
```

This does not give any warning about the duplicate registration. I don't think it comes from within the setupCustomerizers() since that is for the default ones. Unless you've somehow overridden that? 

By default it tries to load:

```
org.openscience.cdk.libio.cml.MDMoleculeCustomizer
org.openscience.cdk.libio.cml.PDBAtomCustomizer
org.openscience.cdk.libio.cml.QSARCustomizer
```

Will do a little more digging but if you have a minimal self contained example it would help. 

Sorry for delay - clearing the back log this week.

**Comment by johnmay** [comment]
Okay I see now :-), the logic was wrong.

```
if (customizers.containsKey(customizerName)) {
   // load
} else {
   // dupe
}
```

should be

```
if (!customizers.containsKey(customizerName)) {
   // load
} else {
   // dupe
}
```

**Comment by johnmay** [comment]
Fixed via 3424cf23909aa36fac01b43ee4658b473b77dd9d - accidentally push right to master.

**Comment by egonw** [comment]
Retrospectively approved! Thanks for the catch! Couldn't find the time to debug it, and slightly embarrassed the problem was this still and trivial :(
